crypto/kdf/sshkdf.c

```diff
@@ -0,0 +1,287 @@
+    size_t len;
+
+    p = va_arg(args, const unsigned char *);
+    len = va_arg(args, size_t);
+    OPENSSL_clear_free(*dst, *dst_len);
+    *dst = OPENSSL_memdup(p, len);
+    if (*dst == NULL)
+        return 0;
+
+    *dst_len = len;
+    return 1;
+}
+
+static int kdf_sshkdf_ctrl(EVP_KDF_IMPL *impl, int cmd, va_list args)
+{
+    int t;
+
+    switch (cmd) {
+    case EVP_KDF_CTRL_SET_MD:
+        impl->md = va_arg(args, const EVP_MD *);
+        if (impl->md == NULL)
+            return 0;
+
+        return 1;
+
+    case EVP_KDF_CTRL_SET_KEY:
+        return kdf_sshkdf_parse_buffer_arg(&impl->key,
+                                           &impl->key_len, args);
+
+    case EVP_KDF_CTRL_SET_SSHKDF_XCGHASH:
+        return kdf_sshkdf_parse_buffer_arg(&impl->xcghash,
+                                           &impl->xcghash_len, args);
+
+    case EVP_KDF_CTRL_SET_SSHKDF_SESSION_ID:
+        return kdf_sshkdf_parse_buffer_arg(&impl->session_id,
+                                           &impl->session_id_len, args);
+
+    case EVP_KDF_CTRL_SET_SSHKDF_TYPE:
+        t = va_arg(args, int);
+        if (t < 65 || t > 70) {
+            KDFerr(KDF_F_KDF_SSHKDF_CTRL, KDF_R_VALUE_ERROR);
+            return 0;
+        }
+
+        impl->type = (char)t;
+        return 1;
+
+    default:
+        return -2;
+
+    }
+}
+
+static int kdf_sshkdf_ctrl_str(EVP_KDF_IMPL *impl, const char *type,
+                               const char *value)
+{
+    if (value == NULL) {
+        KDFerr(KDF_F_KDF_SSHKDF_CTRL_STR, KDF_R_VALUE_MISSING);
+        return 0;
+    }
+
+    if (strcmp(type, "md") == 0)
+        return kdf_md2ctrl(impl, kdf_sshkdf_ctrl, EVP_KDF_CTRL_SET_MD, value);
+
+    if (strcmp(type, "key") == 0)
+        return kdf_str2ctrl(impl, kdf_sshkdf_ctrl,
+                            EVP_KDF_CTRL_SET_KEY, value);
+
+    if (strcmp(type, "hexkey") == 0)
+        return kdf_hex2ctrl(impl, kdf_sshkdf_ctrl,
+                            EVP_KDF_CTRL_SET_KEY, value);
+
+    if (strcmp(type, "xcghash") == 0)
+        return kdf_str2ctrl(impl, kdf_sshkdf_ctrl,
+                            EVP_KDF_CTRL_SET_SSHKDF_XCGHASH, value);
+
+    if (strcmp(type, "hexxcghash") == 0)
+        return kdf_hex2ctrl(impl, kdf_sshkdf_ctrl,
+                            EVP_KDF_CTRL_SET_SSHKDF_XCGHASH, value);
+
+    if (strcmp(type, "session_id") == 0)
+        return kdf_str2ctrl(impl, kdf_sshkdf_ctrl,
+                            EVP_KDF_CTRL_SET_SSHKDF_SESSION_ID, value);
+
+    if (strcmp(type, "hexsession_id") == 0)
+        return kdf_hex2ctrl(impl, kdf_sshkdf_ctrl,
+                            EVP_KDF_CTRL_SET_SSHKDF_SESSION_ID, value);
+
+    if (strcmp(type, "type") == 0) {
+        if (strlen(value) != 1) {
+            KDFerr(KDF_F_KDF_SSHKDF_CTRL_STR, KDF_R_VALUE_ERROR);
+            return 0;
+        }
+
+        return call_ctrl(kdf_sshkdf_ctrl, impl, EVP_KDF_CTRL_SET_SSHKDF_TYPE,
+                         (int)value[0]);
```

Comment:
I think this probably needs to be converted if CHARSET_EBCDIC is defined

Reply:
In a previous comment: https://github.com/openssl/openssl/pull/7290#discussion_r219705508 @richsalz suggest we go the other way, as I was initially using chars.
I am more convinced by @richsalz's comment, as it would be rather hard to know what charset is used as input. So The input, although it uses a char, is indeed documented to be byte value and I provide macros that are charset independent to further help users of the API to pass in the correct values, see  include/openssl/kdf.h for the macros starting with: EVP_KDF_SSHKDF_TYPE_ININITAL_IV_CLI_TO_SRV